File: cpp/src/interfaz.cpp
#include "../include/interfaz.h"
#include <iostream>
#include <cctype>
#include <limits>

#ifdef _WIN32
    #include <conio.h>
    #include <windows.h>
    #include <io.h>
#else
    #include <termios.h>
    #include <unistd.h>
#endif

using namespace std;

void habilitarColoresTerminal() {
#ifdef _WIN32
    HANDLE hOut = GetStdHandle(STD_OUTPUT_HANDLE);
    if (hOut != INVALID_HANDLE_VALUE) {
        DWORD dwMode = 0;
        if (GetConsoleMode(hOut, &dwMode)) {
            dwMode |= ENABLE_VIRTUAL_TERMINAL_PROCESSING;
            SetConsoleMode(hOut, dwMode);
        }
    }
#endif
}

void limpiarPantalla() {
#ifdef _WIN32
    system("cls");
#else
    system("clear");
#endif
}

void encabezadoPrincipal() {
    cout << Colores::CYAN << Colores::BOLD
         << "============================================================\n"
         << "                    PARCIAL 1\n"
         << "              SISTEMA UNIVERSITARIO\n"
         << "============================================================\n"
         << Colores::RESET;
}

void tituloSeccion(const string& titulo) {
    cout << "\n" << Colores::CYAN << Colores::BOLD
         << "--- " << titulo << " ---\n"
         << Colores::RESET;
}

void separador(int ancho) {
    cout << Colores::DIM << string(ancho, '-') << Colores::RESET << "\n";
}

void mensajeExito(const string& msg) {
    cout << Colores::GREEN << Colores::BOLD << "[OK] " << Colores::RESET << Colores::GREEN << msg << Colores::RESET << "\n";
}

void mensajeError(const string& msg) {
    cout << Colores::RED << Colores::BOLD << "[ERROR] " << Colores::RESET << Colores::RED << msg << Colores::RESET << "\n";
}

void mensajeInfo(const string& msg) {
    cout << Colores::CYAN << "[i] " << Colores::RESET << msg << "\n";
}

void mensajeAlerta(const string& msg) {
    cout << Colores::YELLOW << Colores::BOLD << "[!] " << Colores::RESET << Colores::YELLOW << msg << Colores::RESET << "\n";
}

// Lee UN SOLO caracter sin esperar ENTER. Es una funcion "privada" de
// este archivo (static): solo la usan las funciones publicas de aqui abajo.
static char leerCaracterInmediato() {
#ifdef _WIN32
    if (_isatty(_fileno(stdin))) {
        char c = _getch();
        cout << c;
        return c;
    } else {
        char c = '\n';
        if (cin.get(c)) {
            return c;
        }
        return '\n';
    }
#else
    char c;
    termios configuracionOriginal, configuracionNueva;

    tcgetattr(STDIN_FILENO, &configuracionOriginal); // guardar como estaba la terminal
    configuracionNueva = configuracionOriginal;
    configuracionNueva.c_lflag &= ~ICANON;            // desactivar "modo canonico"
    // (modo canonico = la terminal espera ENTER antes de entregar la entrada
    //  al programa; al desactivarlo, cada tecla llega de inmediato)
    tcsetattr(STDIN_FILENO, TCSANOW, &configuracionNueva);

    c = getchar(); // lee una sola tecla, sin esperar ENTER

    // Si el usuario presiono una tecla extra por costumbre (ej. ENTER
    // despues de "s"), esa tecla queda esperando sin usarse. Se descarta
    // aqui para que no interfiera con la siguiente lectura del programa.
    tcflush(STDIN_FILENO, TCIFLUSH);

    tcsetattr(STDIN_FILENO, TCSANOW, &configuracionOriginal); // restaurar la terminal
    return c;
#endif
}

void pausar() {
    cout << "\nPresione una tecla para continuar...";
    leerCaracterInmediato(); // NO importa cual tecla sea, cualquiera continua
    cout << "\n";
}

char leerOpcionMenu(const string& mensaje) {
    char c;
    while (true) {
        cout << mensaje;
        c = leerCaracterInmediato();
        cout << "\n";
        if (c == '\n' || c == '\r') {
            return '0';
        }
        if (isalnum(static_cast<unsigned char>(c))) {
            return static_cast<char>(tolower(static_cast<unsigned char>(c)));
        }
        cout << "Opcion invalida. Ingrese una de las opciones mostradas.\n";
    }
}

int leerOpcionInmediata(const string& mensaje) {
    char c;
    while (true) {
        cout << mensaje;
        c = leerCaracterInmediato();
        cout << "\n";
        if (c == '\n' || c == '\r') {
            // ENTER por si solo se interpreta como "Volver"/"Salir" (opcion 0).
            // Es el comportamiento que la mayoria espera por instinto.
            return 0;
        }
        if (isdigit(static_cast<unsigned char>(c))) {
            return c - '0'; // convierte el caracter '0'-'9' al numero real
        }
        cout << "Opcion invalida. Ingrese un numero de las opciones mostradas.\n";
    }
}

char leerSiNo(const string& mensaje) {
    char c;
    while (true) {
        cout << mensaje;
        c = static_cast<char>(tolower(leerCaracterInmediato()));
        cout << "\n";
        if (c == 's' || c == 'n') return c;
        cout << "Opcion invalida. Por favor ingrese 's' o 'n'.\n";
    }
}

string leerPalabra(const string& mensaje) {
    string palabra;
    cout << mensaje;
    cin >> palabra;
    // Deja el buffer limpio (sin '\n' pendiente) para que un getline()
    // que venga despues no se salte una linea por error.
    cin.ignore(numeric_limits<streamsize>::max(), '\n');
    return palabra;
}

bool esCancelar(const string& texto) {
    string copia = texto;
    for (auto& c : copia) c = static_cast<char>(tolower(static_cast<unsigned char>(c)));
    return copia == "cancelar";
}

int leerEntero() {
    int valor;
    while (true) {
        cin >> valor;
        if (cin.fail()) {
            cin.clear();
            cin.ignore(numeric_limits<streamsize>::max(), '\n');
            cout << "Entrada invalida. Ingrese un numero entero: ";
            continue;
        }
        cin.ignore(numeric_limits<streamsize>::max(), '\n');
        return valor;
    }
}

float leerFlotante() {
    float valor;
    while (true) {
        cin >> valor;
        if (cin.fail()) {
            cin.clear();
            cin.ignore(numeric_limits<streamsize>::max(), '\n');
            cout << "Entrada invalida. Ingrese una nota valida (ej. 3.5): ";
            continue;
        }
        cin.ignore(numeric_limits<streamsize>::max(), '\n');
        return valor;
    }
}

File: python/src/interfaz.py
"""Interfaz de consola: limpiar pantalla, pausas, lectura de opciones sin
ENTER, confirmaciones s/n. Es el equivalente en Python de cpp/src/interfaz.cpp
-- mismas decisiones de diseno, mismo comportamiento para el usuario."""

import os
import sys

# La consola NO se controla igual en todos los sistemas operativos.
# Windows tiene su propia libreria (msvcrt) para leer teclas sueltas;
# Linux/Mac usan la libreria estandar termios/tty para lograr lo mismo.
# Se detecta automaticamente cual esta disponible al importar el modulo.
try:
    import msvcrt
    _ES_WINDOWS = True
except ImportError:
    import termios
    import tty
    _ES_WINDOWS = False


class Colores:
    RESET   = "\033[0m"
    BOLD    = "\033[1m"
    DIM     = "\033[2m"
    CYAN    = "\033[36m"
    GREEN   = "\033[32m"
    YELLOW  = "\033[33m"
    RED     = "\033[31m"
    BLUE    = "\033[34m"
    MAGENTA = "\033[35m"


def limpiar_pantalla():
    os.system("cls" if os.name == "nt" else "clear")


def encabezado_principal():
    print(f"{Colores.CYAN}{Colores.BOLD}"
          f"============================================================\n"
          f"                    PARCIAL 1\n"
          f"              SISTEMA UNIVERSITARIO\n"
          f"============================================================{Colores.RESET}")


def titulo_seccion(titulo):
    print(f"\n{Colores.CYAN}{Colores.BOLD}--- {titulo} ---{Colores.RESET}")


def separador(ancho=60):
    print(f"{Colores.DIM}{'-' * ancho}{Colores.RESET}")


def mensaje_exito(msg):
    print(f"{Colores.GREEN}{Colores.BOLD}[OK] {Colores.RESET}{Colores.GREEN}{msg}{Colores.RESET}")


def mensaje_error(msg):
    print(f"{Colores.RED}{Colores.BOLD}[ERROR] {Colores.RESET}{Colores.RED}{msg}{Colores.RESET}")


def mensaje_info(msg):
    print(f"{Colores.CYAN}[i] {Colores.RESET}{msg}")


def mensaje_alerta(msg):
    print(f"{Colores.YELLOW}{Colores.BOLD}[!] {Colores.RESET}{Colores.YELLOW}{msg}{Colores.RESET}")


def _leer_caracter_inmediato():
    """Lee UN SOLO caracter sin esperar ENTER. Funcion 'privada' de este
    modulo (empieza con _): solo la usan las funciones publicas de aqui abajo."""
    if _ES_WINDOWS and sys.stdin.isatty():
        c = msvcrt.getch().decode("utf-8", errors="ignore")
        print(c, end="", flush=True)  # msvcrt no muestra el caracter, lo mostramos nosotros
        return c
    elif _ES_WINDOWS:
        ch = sys.stdin.read(1)
        return ch if ch else "\n"
    else:
        fd = sys.stdin.fileno()
        try:
            configuracion_original = termios.tcgetattr(fd)
        except termios.error:
            # No hay una terminal real disponible (ej. entrada redirigida
            # desde un archivo, IDE que no da consola interactiva real).
            # Respaldo: leer una linea normal en vez de caerse.
            linea = sys.stdin.readline()
            return linea[0] if linea else "\n"
        try:
            tty.setcbreak(fd)  # desactiva el "modo canonico" (no espera ENTER)
            c = sys.stdin.read(1)
        finally:
            termios.tcsetattr(fd, termios.TCSADRAIN, configuracion_original)
        return c


def pausar():
    """Espera a que el usuario presione CUALQUIER tecla para continuar."""
    print("\nPresione una tecla para continuar...", end="", flush=True)
    _leer_caracter_inmediato()
    print()


def leer_opcion_menu(mensaje):
    """Lee un caracter de opcion de menu alfanumerico (ej: 0-9, a, r)."""
    while True:
        print(mensaje, end="", flush=True)
        c = _leer_caracter_inmediato()
        print()
        if c in ("\n", "\r", ""):
            return "0"
        if c.isalnum():
            return c.lower()
        print("Opcion invalida. Ingrese una de las opciones mostradas.")


def leer_opcion_inmediata(mensaje):
    """Lee UN SOLO digito (0-9) como opcion de menu, sin esperar ENTER.
    ENTER por si solo se interpreta como 0 (Volver/Salir) — comportamiento
    esperado por instinto."""
    while True:
        print(mensaje, end="", flush=True)
        c = _leer_caracter_inmediato()
        print()
        if c in ("\n", "\r", ""):
            return 0
        if c.isdigit():
            return int(c)
        print("Opcion invalida. Ingrese un numero de las opciones mostradas.")


def leer_si_no(mensaje):
    """Pregunta algo tipo '(s/n)' y lee UN SOLO caracter, sin esperar ENTER.
    Si no es 's' ni 'n', avisa y vuelve a preguntar."""
    while True:
        print(mensaje, end="", flush=True)
        c = _leer_caracter_inmediato().lower()
        print()
        if c in ("s", "n"):
            return c
        print("Opcion invalida. Por favor ingrese 's' o 'n'.")


def leer_palabra(mensaje):
    """Lee una palabra normal (con ENTER, como cualquier input de texto)."""
    return input(mensaje).strip()


def es_cancelar(texto):
    """True si el texto (sin importar mayusculas/minusculas) es 'cancelar'.
    Se usa junto con leer_palabra() para poder abortar una operacion
    escribiendo esa palabra en vez de un codigo."""
    return texto.strip().lower() == "cancelar"


def leer_entero(mensaje=""):
    """Lee un numero entero de forma segura. Si el usuario escribe texto
    en vez de un numero, no se cuelga: avisa y vuelve a pedirlo."""
    while True:
        try:
            return int(input(mensaje).strip())
        except ValueError:
            print("Entrada invalida. Ingrese un numero.")
